### PR TITLE
31 resource collection

### LINKED ISSUE #31: Resource Collection
## Description
As a player, I want my pawns to be able to collect the resources in the world, in order to amass resources.
## Requirements
- [x] Roles create actions
- [x] Beings perform actions
- [x] The pawns must go up to resources to collect them (without destroying them)
- [x] The pawns needs to find the nearest resource
- [x] A pawn only collects resources from what their role is assigned
- [x] #67

## Dependencies
#11

### PR DESCRIPTION
**Pull request by martinjonsson01**
## Summary
hPQ1drRXBe
The action system has been added. Roles now give out Actions to Beings, which Beings then perform. 

Beings with the appropriate roles now collect the resources: 
* Fisher          harvests Water to get Fish
* Lumberjack harvests Tree to get Log
* Miner          harvests Rock to get Stone

When a being is not able to perform their role, they wander around as if they were idle.

## Test Coverage
image
image
image
image
image

## Definition of Done
- [x] Completed all Tasks of the User Story
- [x] The Tasks have tests and they all pass
- [x] Overall test coverage is >= 90%
- [x] A UML class diagram of the newly implemented User Story is done and included in the README.md of the package

## Diagrams
com thebois models beings actions
com thebois models beings roles
com thebois
com thebois models beings
com thebois abstractions
com thebois models world
com thebois models world resources

Closes #31

### CODE REVIEW

File: config/suppressions.xml
Comment (jonathan-lindqvist) [inline config/suppressions.xml:9]: Oh no

File: src/main/java/com/thebois/abstractions/README.md
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/abstractions/README.md:1]: What?

File: src/main/java/com/thebois/models/beings/AbstractBeing.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/AbstractBeing.java:20]: Question

Guessing we use this to see if we have reached our destination, would make sense. But do we REALLY need it here?

File: src/main/java/com/thebois/models/beings/AbstractBeing.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/AbstractBeing.java:56]: 🥳

File: src/main/java/com/thebois/models/beings/Colony.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/Colony.java:36]: `this`?

File: src/main/java/com/thebois/models/beings/Pawn.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/Pawn.java:15]: Gut 💯

File: src/main/java/com/thebois/models/beings/actions/DoNothingAction.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/actions/DoNothingAction.java:24]: Just curious, is this if they are idle?

File: src/main/java/com/thebois/models/beings/actions/IActionGenerator.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/actions/IActionGenerator.java:6]: I mean, its the name of the interface :P

File: src/main/java/com/thebois/models/beings/actions/MoveAction.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/actions/MoveAction.java:53]: Do we need this?

File: src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:18]: Mainly?

File: config/suppressions.xml
Comment (martinjonsson01) [reply config/suppressions.xml:9]: Oh yes

File: src/main/java/com/thebois/models/beings/AbstractBeing.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeing.java:20]: It's not *the* actual final destination, it's the next tile that the being is moving towards. This step-towards-the-final-destination is the being's temporary destination, given to it from the `MoveAction` which holds the *actual* final destination.

File: src/main/java/com/thebois/models/beings/Colony.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/Colony.java:36]: `Colony` implements `IInventory` and the `Pawn` needs access to the colony inventory.

File: src/main/java/com/thebois/models/beings/actions/DoNothingAction.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/actions/DoNothingAction.java:24]: Nope, when they're idle, they perform different `MoveAction`s with random destinations. This action is used when a `Role` wants to perform an actual `Action` but can't, for some reason (e.g. the tree it is trying to get to is blocked on all four sides). 

For how idle is handled, see `IdleRole`, and for how `DoNothingAction` is used, see `AbstractHarvestRole`.

File: src/main/java/com/thebois/models/beings/actions/IActionGenerator.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/actions/IActionGenerator.java:6]: Yeah, but that's all it is, both in name and in function. Sometimes the name is different than the function, like for `Being`, where the function is that it's an entity that can do things in the world. For this, though, which is actually just a very simple representation of a lambda method, I chose to name it exactly after its function, so they line up.

Couldn't come up with anything more fitting, maybe you can?

Any suggestions for another name, or another way to describe its functionality?

File: src/main/java/com/thebois/models/beings/actions/MoveAction.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/actions/MoveAction.java:53]: For debugging tests, yes, 100%. It's very difficult otherwise. Just like with `Position` (which you got to see earlier today).

**Review by Matte02 — CHANGES_REQUESTED** [review]
Doing the same as Jonathan, this was what I found while skimming over the code. 

Overall really good, some small questions and requests. 

Will try to do a more thorough review tomorrow.

File: src/main/java/com/thebois/models/world/World.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:226]: Yeah, I intentionally implemented it in the most simple way (aka dumbest way) I could think of. That's because it is also the most readable, and easy to understand way. Like with the pathfinding, I chose not to preemptively optimize.

When running the game in a 50x50 world, there aren't any problems. And this isn't a method that is called every frame, it's called much more seldom than that. 

What do you think, should we be benchmarking the game with a bigger world? Or should we optimize for the 50x50 size and be happy with if that works? To be honest, I'm not sure, we might want to up the size, but then the question is where we draw the line. Any ideas?

File: src/main/java/com/thebois/models/world/resources/Stone.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/resources/Stone.java:17]: Oh you're right.

Really, the javadoc shouldn't have to mention the name of the class. It's already apparent that the javadoc is describing the class, because its located inside of the class. Here we see a great example of why to avoid writing out type names in javadoc.

I'll fix it.

Comment (Matte02) [reply src/main/java/com/thebois/models/world/resources/Stone.java:17]: Sorry

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/resources/Stone.java:17]: It's no problem, it's a very easy mistake to make! It's more natural to write it the way you have, so you kind of have to force yourself to leave out the "for xxxx" at the end.

File: src/main/java/com/thebois/models/world/World.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:226]: Good idea, where do you think mentioning that would fit in the SDD?

File: src/main/java/com/thebois/models/world/World.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:285]: It took a bit of fiddling with the tests, but now pawns spawn in random locations!

File: src/main/java/com/thebois/models/world/World.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:228]: Maybe known issues?

Comment (Matte02) [reply src/main/java/com/thebois/models/world/World.java:228]: That a good spot for it, since it is a know issue

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:228]: Added 👍

File: src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java
Comment (Matte02) [inline src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:67]: Tree?
This is the AbstractHarvesterRole, pretty sure it should be resource, maybeResource etc.

File: src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java
Comment (Matte02) [inline src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:60]: I might have missed something, but why are these methods protected. The only method I can see being used by subclasses are getTaskGenerators but not the other 3 methods.

File: src/main/java/com/thebois/models/beings/roles/RoleFactory.java
Comment (Matte02) [inline src/main/java/com/thebois/models/beings/roles/RoleFactory.java:136]: Question
This method is being called only to make sure that every role has a world and finder, correct? 
Mostly asking because I've never seen a method like this before.

File: src/main/java/com/thebois/models/world/IWorld.java
Comment (Matte02) [inline src/main/java/com/thebois/models/world/IWorld.java:55]: Do we want to mention that if no neighbour was vacant, then the optional will be empty or is that understood by having an optional? 
Cause that is what I'm assuming the method should do, but from reading the Javadoc I feel like it returns the closest neighbour even if it's occupied.

What I'm trying to say is that we should mention in the Javadoc that it will give the closestVacantNeighbour.

File: src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:67]: Good catch, I missed that when I refactored lumberjack into the harvest role!

File: src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:60]: You're right, they're not used in any subtype as of now.

I made them protected because I can imagine that some harvesting-roles will want to define different ways of finding resources, and different ways of harvesting them. But maybe that assumption is wrong, what do you think?

The argument could be made that this is speculative generality, and that they should only be made protected when they actually need to be, not before.

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:60]: The downside to making them private now is that whenever some hypothetical future programmer wants to implement a unique harvesting role that finds resources only by wandering around (instead of instantly knowing where they are) and that only harvests them by dancing around in a circle first, they'll have to go way up into `AbstractHarvesterRole` and change the visibility of these methods.

My reasoning is that, since it is an abstract class, it's good that it declares some methods as protected, so subtypes can access/override them.

File: src/main/java/com/thebois/models/world/IWorld.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/IWorld.java:55]: You're right, the fact that it returns an `Optional<Position>` implies that it might not return anything. But you're also right that it's good to be really clear in the javadoc about that behaviour as well, so the reader doesn't have to infer anything. 

Updating it.

File: src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java
Comment (Matte02) [reply src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:60]: Hmm, I can understand your way of thinking, but I still feel like they should be private until they need to be protected. 
What do you think @jonathan-lindqvist  and @JacobBredin?

File: src/main/java/com/thebois/models/beings/AbstractBeing.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/AbstractBeing.java:20]: 👍

File: src/main/java/com/thebois/models/beings/Colony.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/Colony.java:36]: There is a `getInventory` in colony, couldn't Actions use that and only give it to the pawn when interactions are needed?

File: src/main/java/com/thebois/models/beings/actions/DoNothingAction.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/actions/DoNothingAction.java]: 👍

File: src/main/java/com/thebois/models/beings/actions/IActionGenerator.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/actions/IActionGenerator.java:6]: Yeah that it better, and more consistent with the rest of the project 👍

File: src/main/java/com/thebois/models/beings/actions/MoveAction.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/actions/MoveAction.java:53]: But debugging code in "production" 😬

File: src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:18]: 👍

File: src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:60]: I agree with @Matte02 
They should be private, taking the _hypothetical future programmer_ argument to the unnecessary extreme, we could make everything public because someone in the future might want to access this.

File: src/main/java/com/thebois/models/beings/roles/GuardRole.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/roles/GuardRole.java:19]: once again i lost, could you breakdown this line?

File: src/main/java/com/thebois/models/beings/roles/IdleRole.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/roles/IdleRole.java:38]: Does `World` have a `getRandomVacantSpot()`? It's not wrong but feels sort of weird as well, don't you agree? The other way would be either the Action or Pathfinder would generate a random pos and ask if its vacant. But might not be as elegant in reality as it sounds in my head atm. What is your opinion?

File: src/main/java/com/thebois/models/beings/Colony.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/Colony.java:36]: The `HarvestAction` only has access to an `IActionPerformer`. In `IActionPerformer`, there is an `addItem` method. Since it is `AbstractBeing` which implements `IActionPerformer`, it has to implement `addItem`. Since `AbstractBeing` doesn't have an inventory of its own, it needs the inventory injected from `Colony`.

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/Colony.java:36]: But, once we've merged #83 into master, `AbstractBeing` will have its own inventory. So it'll no longer need to have it injected from `Colony`.

File: src/main/java/com/thebois/models/world/World.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/world/World.java:45]: Will we use a different `Random`? Because the only thing we do is create it in pawntastic and immediately send it to world

File: src/main/java/com/thebois/models/beings/actions/IActionGenerator.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/actions/IActionGenerator.java:6]: Fixing, but I don't see how it's more consistent? Do we have other `IxxxxSource` interfaces?

File: src/main/java/com/thebois/views/debug/BeingPathDebugView.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/debug/BeingPathDebugView.java:19]: Does _the view_ recalculate the paths?

File: src/main/java/com/thebois/models/beings/actions/MoveAction.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/actions/MoveAction.java:53]: It's **toString**, one of the fundemental methods built into all objects in Java. It's not weird to have an object implementing it, no matter the reason.

Representing an object in text format is something intrinsic to all objects, like how you're able to print an `int` even though it is not made out of text. In the same way, you can't really print `MoveAction` because it is not made out of text. But, like with `int`, you can represent it using text, which is what this override does. 

>debugging code in "production"

This is like not wanting analytics in production because it's not relevant to making the application function. It's an auxiliary function of the application. In the case of analytics, it makes it easier to track behavior and find errors, and in the case of `toString` it makes it easier to find errors, by letting other developers understand why they broke the tests. 

It's like not wanting catch-blocks to print the error because that's debug-code.

So what I'm really trying to argue is that having a string representation of `MoveAction` is a needed functionality, because `MoveAction` is a value object, and when some other developer later on changes some code that breaks `MoveAction`-related code, they'll want error messages saying
`"Expected MoveAction{destination={x=1, y=2}}, but actual is MoveAction{destination={x=10, y=20}}"`
instead of what they would get by default: 
`"Expected MoveAction@2314 but actual is MoveAction@2315"`.
The latter tells the programmer nothing at all, which makes it much harder to debug, while the former makes it incredibly easy to see at a glance what has gone wrong. This is auxiliary functionality that is useful in the same way having tests are.

You wouldn't call having unit tests "debugging in production", would you? I'd argue that this `toString` is just like the unit tests.

File: src/main/java/com/thebois/models/beings/roles/GuardRole.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/GuardRole.java:19]: Sure, so `GuardRole` does not have any unique functionality implemented yet. So the behaviour we want is for it to act like an idle being, just randomly moving around the world.

The `getTaskGenerators` method is how different roles provide different sets of action-generators, and these generators then return the actions themselves. In `AbstractRole`, inside `obtainNextAction`, `getTaskGenerators` (the template method) is called when trying to obtain the next action to perform. For `GuardRole`, it comes to this method on line 19.

Here, a new idle role is created from the `RoleFactory`, and then the responsibility of generating `Action`s is delegated to that role. So, in effect, making `GuardRole` behave as if it was an `IdleRole`.

File: src/main/java/com/thebois/models/beings/roles/IdleRole.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/IdleRole.java:38]: It was the way you describe, before. 

The reason I changed it is because the responsibility of finding an empty position in the `World` is not `IdleRole`'s. It's `World`'s. Why? It's easy to notice when you write out how it would look if it weren't done in `World`:
```java
Position randomPos;
do{
    randomPos = generateRandomPosition();
} while(!world.isVacant(randomPos));
```
It's a violation of the law of demeter. `IdleRole` shouldn't repeatedly query the world, it should just ask the world for what the world knows. Since the world knows everything about what is inside of it, then only world should do the work of finding a random vacant spot.

**Comment by martinjonsson01** [comment]
>would really like you to go trough the code with me IRL

Sure! We can do it this Tuesday.

>should we remove all roles that we wont have time to give a purpose

Probably a good idea, since they just act like idle beings right now.

File: src/main/java/com/thebois/views/debug/BeingPathDebugView.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/debug/BeingPathDebugView.java:19]: To elaborate, now that beings themselves no longer are in charge of or even know about the path that they're taking, this debug view can no longer access it. It's hidden deep within the `MoveAction`, which is entirely inaccessible from the outside (as it should be). 

This means it's basically impossible to render the path they're planning on walking, unless there is some way to get the final destination they're moving towards, and then calculating the path again. Since the beings themselves don't know about this final destination either, the best they can give is their current next-tile-destination. The debug view then calculates the path to that destination and renders it, since it can't get the real path from inside the being.

File: src/main/java/com/thebois/models/beings/actions/MoveAction.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/actions/MoveAction.java:57]: 👍

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/actions/MoveAction.java:57]: Too annoying to argue, eh? 😅 Sorry about that

File: src/main/java/com/thebois/models/beings/roles/GuardRole.java
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/roles/GuardRole.java:19]: 👍

**Review by JacobBredin — CHANGES_REQUESTED** [review]

File: src/main/java/com/thebois/models/beings/AbstractBeing.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeing.java:89]: Good idea, I'll see if I can get around to that (will focus on this PR for now). If you feel eager to fix it, then you can do it as a hotfix if you'd like!

File: src/main/java/com/thebois/models/beings/actions/ActionFactory.java
Comment (JacobBredin) [inline src/main/java/com/thebois/models/beings/actions/ActionFactory.java:42]: 👍

File: src/main/java/com/thebois/models/beings/roles/IdleRole.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/roles/IdleRole.java:21]: Doesn't seem so random to me

File: src/main/java/com/thebois/views/game/ColonyView.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/game/ColonyView.java:32]: 👍

File: src/main/java/com/thebois/views/game/ColonyView.java
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/game/ColonyView.java:75]: Is this more readable?

**Review by JacobBredin — APPROVED** [review]
Lgtm, well done! One small comment to address first

File: src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:45]: Hm, could you elaborate on how it's confusing?

As you say, adding "Action" to the end doesn't clarify anything, since that's the return type already.

The way I read it right now is "creates a move-to-resource (action)", did you read it in a different way, maybe?

File: src/main/java/com/thebois/models/beings/Colony.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/Colony.java:52]: I got tired of the unreadable `something.doThing().performthingy(withThisThing).whenThereIsABigPointyTriangle(inTheColorBlue).doADanceAndSingThisSong(SongFactory.createTralalalala());` so i changed the formatter to wrap chained method calls.

I admit, that in this spot it is very redundant. But on the whole, I think it makes long method chains a lot more readable. Or do you think that the downsides outweigh the benefits?

File: src/main/java/com/thebois/models/beings/roles/IdleRole.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/IdleRole.java:21]: True, the radius itself isn't what's random. It's the radius in which random movement locations are picked. Do you have any suggestions for other names?

File: src/main/java/com/thebois/Pawntastic.java
Comment (martinjonsson01) [reply src/main/java/com/thebois/Pawntastic.java:201]: Ah, I see.

Yeah, since the factories don't get saved to the save file. The factories are instantiated from scratch again when the program is closed and opened, and then their static fields are null again. So they need to be called to be re-set.

Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/Pawntastic.java]: We will see how that works out for stockpiles then 😬

**Pull request merged**
